src/tasks.jl

```diff
@@ -0,0 +1,233 @@
+    end
+    res.task.storage[:turing_trace] = res # create a backward reference in task_local_storage
+    return res
+end
+
+
+
+# step to the next observe statement, return log likelihood
+Libtask.consume(t::Trace) = (t.vi.num_produce += 1; consume(t.task))
+
+# Task copying version of fork for Trace.
+function fork(trace :: Trace, copy_vi::Function, is_ref :: Bool = false, set_retained_vns_del_by_spl!::Union{Function,Nothing} = nothing)
+    newtrace = copy(trace,copy_vi)
+    if is_ref
+        @assert set_retained_vns_del_by_spl! != nothing "[AdvancedPF] set_retained_vns_del_by_spl! is not set."
+        set_retained_vns_del_by_spl!(newtrace.vi)
+    end
+    newtrace.task.storage[:turing_trace] = newtrace
+    return newtrace
+end
+
+# PG requires keeping all randomness for the reference particle
+# Create new task and copy randomness
+function forkr(trace :: Trace, copy_vi::Function)
+
+    newtrace = Trace(trace.vi,trace.task ,trace.taskinfo,copy_vi)
+    newtrace.vi.num_produce = 0
+    return newtrace
+end
+
+current_trace() = current_task().storage[:turing_trace]
+
+const Particle = Trace
+
+"""
+Data structure for particle filters
+- effectiveSampleSize(pc :: ParticleContainer)
+- normalise!(pc::ParticleContainer)
+- consume(pc::ParticleContainer): return incremental likelihood
+"""
+mutable struct ParticleContainer{Tvi,TInfo} <: AbstractParticleContainer where {Tvi, TInfo <: AbstractTaskInfo}
+    vals::Vector{Trace{Tvi,TInfo}}
+    # A named tuple with functions to manipulate the particles vi.
+    manipulators::Dict{String,Function}
+    # logarithmic weights (Trace) or incremental log-likelihoods (ParticleContainer)
+    logWs::Vector{Float64}
+    #This corresponds to log p(x_{0:t}), therefore the log likelihood of the transitions up to this point.
+    #Especially important for ancestor sampling.
+    logpseq::Vector{Float64}
+    # log model evidence
+    logE::Float64
+    # helpful for rejuvenation steps, e.g. in SMC2
+    n_consume::Int
+
+
+end
+
+## Empty initilaizaiton
+ParticleContainer{Tvi,TInfo}() where {Tvi, TInfo <: AbstractTaskInfo} = ParticleContainer{Tvi,TInfo}(Dict{String,Function}(),0)
+ParticleContainer{Tvi,TInfo}(manipulators::Dict{String,Function}) where {Tvi, TInfo <: AbstractTaskInfo} = ParticleContainer{Tvi,TInfo}(manipulators,0)
+#Some initilaizaiton...
+function ParticleContainer{Tvi,TInfo}(manipulators::Dict{String,Function}, n::Int) where {Tvi, TInfo <: AbstractTaskInfo}
+    if !("copy" in keys(manipulators)) manipulators["copy"] = deepcopy end
+    if !("set_retained_vns_del_by_spl!" in keys(manipulators)) manipulators["set_retained_vns_del_by_spl!"] = (x) -> () end
+    ParticleContainer{Tvi,TInfo}( Vector{Trace{Tvi,TInfo}}(undef, n), manipulators, Float64[],Float64[], 0.0, 0)
+
+end
+
+
+Base.collect(pc :: ParticleContainer) = pc.vals # prev: Dict, now: Array
+Base.length(pc :: ParticleContainer)  = length(pc.vals)
+Base.similar(pc :: ParticleContainer{T}) where T = ParticleContainer{T}(0)
```

Comment:
That's not really a proper implementation of `similar`, since it doesn't respect the size of the current container. It was not needed before, so I just removed it in Turing.
```suggestion
```

Reply:
This is actually used in the resample.jl file

Reply:
You mean in https://github.com/TuringLang/AdvancedPS.jl/blob/f1f98652324d527c5ab149c6a9d3fbb319fd03f6/src/resample.jl#L18 ? To me it seems there `particles` is actually of type `Vector{<:Trace}` and not of type `ParticleContainer`, so that doesn't call the method we're discussing here. Removing it in Turing also didn't break anything :slightly_smiling_face: